src/main/java/org/opensearch/flowframework/workflow/WorkflowProcessSorter.java

```diff
@@ -74,17 +75,68 @@ public List<ProcessNode> sortProcessNodes(Workflow workflow) {
                 .collect(Collectors.toList());
 
             TimeValue nodeTimeout = parseTimeout(node);
-            ProcessNode processNode = new ProcessNode(node.id(), step, data, predecessorNodes, threadPool, nodeTimeout);
+            ProcessNode processNode = new ProcessNode(
+                node.id(),
+                step,
+                node.previousNodeInputs(),
+                data,
+                predecessorNodes,
+                threadPool,
+                nodeTimeout
+            );
             idToNodeMap.put(processNode.id(), processNode);
             nodes.add(processNode);
         }
 
         return nodes;
     }
 
+    /**
+     * Validates a sorted workflow, determines if each process node's user inputs and predecessor outputs match the expected workflow step inputs
+     * @param processNodes A list of process nodes
+     * @throws Exception on validation failure
+     */
+    public void validateGraph(List<ProcessNode> processNodes) throws Exception {
+
+        WorkflowValidator validator = WorkflowValidator.parse("mappings/workflow-steps.json");
+
+        // Iterate through process nodes in graph
+        for (ProcessNode processNode : processNodes) {
+
+            // Get predecessor nodes types of this processNode
+            List<String> predecessorNodeTypes = processNode.predecessors()
+                .stream()
+                .map(x -> x.workflowStep().getName())
+                .collect(Collectors.toList());
+
+            // Compile a list of outputs from the predecessor nodes based on type
+            List<String> predecessorOutputs = new ArrayList<>();
+            for (String nodeType : predecessorNodeTypes) {
+                List<String> nodeTypeOutputs = validator.getWorkflowStepValidators().get(nodeType).getOutputs();
+                predecessorOutputs.addAll(nodeTypeOutputs);
+            }
+
+            // Retrieve all the user input data from this node
+            List<String> currentNodeUserInputs = new ArrayList<String>(processNode.input().getContent().keySet());
+
+            // Combine both predecessor outputs and current node user inputs
+            List<String> allInputs = new ArrayList<>();
+            allInputs.addAll(predecessorOutputs);
+            allInputs.addAll(currentNodeUserInputs);
+
+            // Retrieve list of required inputs from the current process node and compare
+            List<String> expectedInputs = validator.getWorkflowStepValidators().get(processNode.workflowStep().getName()).getInputs();
+            if (!allInputs.containsAll(expectedInputs)) {
+                expectedInputs.removeAll(allInputs);
+                throw new IllegalArgumentException("Invalid graph, missing the following required inputs : " + expectedInputs.toString());
```

Comment:
We need to throw this error in the execute API with a 4xx.

Reply:
Sure, that makes sense. Currently, graph validation happens **after** we return a response back to the user's `_provision` call, but it makes sense to have validation before we attempt to execute. Ill make this modification so that we can return 4xx on failure to validate

Reply:
Do you mean that it will be part of the create API then?

Reply:
There's an issue https://github.com/opensearch-project/opensearch-ai-flow-framework/issues/99 to add a `dry-run` param to the create API which would do validation on a template there. 

For the provision API, we've also doing validation prior to executing. The modification that I'll make right now is to do validation before we execute the workflow on a separate thread. That way we can return a 4xx response back to the user.

Reply:
@joshpalis I thought a little about this. What do you think of having the validation on create API? Basically #99. Also, that means we will be doing validation on create API only when `dry-run` param is present but for provision API we will be doing it always.

Reply:
Yes, your understanding is correct, we're aligned.